lib/ace/ext/papeeria/tex_completer.coffee

```diff
@@ -75,48 +82,81 @@ define( (require, exports, module) ->
   equationKeywords = ['\\alpha']
   listKeywords = ['\\item']
 
-  listKeywords_ = listKeywords.map((word) ->
+  listKeywords = listKeywords.map((word) ->
     caption: word,
     value: word
     meta: 'list'
   )
-  equationKeywords_ = equationKeywords.map((word) ->
+  equationKeywords = equationKeywords.map((word) ->
     caption: word,
     value: word
     meta: 'equation'
   )
   
-  init = (editor, bindKey) ->
-    HashHandler = require("ace/keyboard/hash_handler").HashHandler; 
-    keyboardHandler = new HashHandler();
+  # Specific for token's system of type in ace
+  # We saw such a realization in html_completions.js
+  istype = (token, type) -> 
+    return token.type.lastIndexOf(type) > -1
+  
+
+  init = (editor, bindKey) -> 
+    HashHandler = require("ace/keyboard/hash_handler").HashHandler
+    keyboardHandler = new HashHandler()
     keyboardHandler.addCommand(
-      name: 'add item in list mode',
-      bindKey: bindKey,
-      exec: (editor) ->
+      name: 'add item in list mode'
+      bindKey: bindKey
+      exec: (editor) ->  
         pos = editor.getCursorPosition()
-        curLine = editor.session.getLine(pos.row);
+        curLine = editor.session.getLine(pos.row)
+        indentCount = ContextHelper.getNestedListDepth(editor.session, pos.row)
+        tabString = editor.getSession().getTabString()
         # it's temporary fix bug with added \item before \begin{itemize|enumerate}
-        if (editor.getSession().getContext(pos.row) == "list" && curLine.indexOf("begin") < pos.column) 
-          editor.insert("\n\t\\item ")
+        if ContextHelper.getContext(editor.session, pos.row) == LIST_STATE && curLine.indexOf("begin") < pos.column
+          editor.insert("\n" + tabString.repeat(indentCount) + "\\item ")
           return true
         else 
           return false
     )
     editor.keyBinding.addKeyboardHandler(keyboardHandler)
 
-  getCompletions = (editor, session, pos, prefix, callback) ->
-    context = session.getContext(pos.row)
-    if context == "start"
-      callback(null, listSnippets.concat(equationSnippets.concat(basicSnippets)))
+   class ReferenceGetter
+    constructor: ->
+      @lastFetchedUrl =  ""
+      @cache = []
+    processData: (data) => 
+      @cache = data.map((elem) => 
+          return {
+            name: elem.caption
+            value: elem.caption
+            meta: elem.type + "-ref" 
+          }
+    )
+    getReferences: (url, callback) => 
+      if url != @lastFetchedUrl
+        $.getJSON(url).done((data) => 
+          @processData(data)
+          callback(null, @cache)
+          @lastFetchedUrl = url
+        )
 
-    if context == LIST_STATE
-      callback(null, listKeywords_.concat(listSnippets.concat(equationSnippets)))
+  class TexCompleter
+      constructor: ->
+        @refGetter = new ReferenceGetter()
+      #callback -- add items in popup
```

Comment:
Please write meaningful documentation. This is not a documentation, this is just a little bit more entropy. Meaningful documentation for a function includes a list of arguments, their type and semantics.